### PR TITLE
Use rows for variant constructors

### PR DESCRIPTION
**Pull request by brendanzab**
Like extensible records, you can't actually do much with these at the moment, but it _does_ simplify the internal representation somewhat.

### CODE REVIEW

File: base/src/types.rs [hunk 1]
```diff
@@ -750,23 +760,31 @@ impl<'a, I, T, E> DisplayType<'a, I, T, E>
...
+                            for arg in arg_iter(&field.typ) {
+                                doc = chain![arena;
+                                            doc,
+                                            " ",
+                                            dt(self.env, Prec::Constructor, &arg).pretty(arena)];
+                            }
+                        }
                     }
-                }
+                    _ => panic!("Polymorphic variant syntax not yet decided"),
```
Comment (Marwes) [inline base/src/types.rs]: "Unexpected type {} in variant" or something? May be a bit clearer about what happened if this is actually reached.

File: base/src/types.rs [hunk 2]
```diff
@@ -903,6 +921,7 @@ pub fn walk_type_<I, T, F: ?Sized>(typ: &T, f: &mut F)
             }
         }
         Type::Record(ref row) => f.walk(row),
+        Type::Variant(ref row) => f.walk(row),
```
Comment (Marwes) [inline base/src/types.rs]: Could use `Type::Record(ref row) | Type::Variant(ref row) =>`

File: check/src/rename.rs [hunk 3]
```diff
@@ -143,9 +143,9 @@ pub fn rename(symbols: &mut SymbolModule,
 
         fn stack_type(&mut self, id: Symbol, span: Span<BytePos>, alias: &Alias<Symbol, ArcType>) {
             // Insert variant constructors into the local scope
-            if let Type::Variants(ref variants) = *alias.typ {
-                for &(ref name, ref typ) in variants {
-                    self.env.stack.insert(name.clone(), (name.clone(), span, typ.clone()));
+            if let Type::Variant(ref row) = *alias.typ {
+                for field in row.field_iter().cloned() {
```
Comment (Marwes) [inline check/src/rename.rs]: Maybe we need to rename `field_iter` now? `row_iter` ?

Comment (brendanzab) [inline check/src/rename.rs]: Want me to change `type_field_iter` as well?

Comment (Marwes) [inline check/src/rename.rs]: That one should not apply to variants so maybe that one is fine.

**Pull request merged**